plugins/vpp/l3plugin/descriptor/vrrp.go

```diff
@@ -0,0 +1,254 @@
+		return kvs.NewInvalidValueError(ErrInvalidInterval, "interval")
+	}
+
+	var ip net.IP
+	var isIpv6 bool
+	for idx, addr := range vrrp.IpAddresses {
+		ip = net.ParseIP(addr)
+		if ip == nil {
+			return kvs.NewInvalidValueError(ErrInvalidVrrpIP, "ip_addresses")
+		}
+
+		if idx == 0 && ip.To4() == nil {
+			isIpv6 = true
+			continue
+		}
+
+		if ip.To4() == nil && !isIpv6 || ip.To4() != nil && isIpv6 {
+			return kvs.NewInvalidValueError(ErrInvalidIPVer, "ip_addresses")
+		}
+	}
+	return nil
+}
+
+// Create adds VPP VRRP entry.
+func (d *VrrpDescriptor) Create(key string, vrrp *l3.VRRPEntry) (interface{}, error) {
+	if err := d.vrrpHandler.VppAddVrrp(vrrp); err != nil {
+		if errors.Is(vppcalls.ErrVRRPUnsupported, err) {
+			d.log.Debugf("Unsupported action: %v", err)
+		}
+		return nil, err
+	}
+
+	if vrrp.Enabled {
+		if err := d.vrrpHandler.VppStartVrrp(vrrp); err != nil {
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+// Delete removes VPP VRRP entry.
+func (d *VrrpDescriptor) Delete(key string, vrrp *l3.VRRPEntry, metadata interface{}) error {
+	if vrrp.Enabled {
+		if err := d.vrrpHandler.VppStopVrrp(vrrp); err != nil {
+			if errors.Is(vppcalls.ErrVRRPUnsupported, err) {
+				d.log.Debugf("Unsupported action: %v", err)
+			}
+			return err
+		}
+	}
+	if err := d.vrrpHandler.VppDelVrrp(vrrp); err != nil {
+		return err
+	}
+	return nil
+}
+
+// UpdateWithRecreate returns true if a VRRP update needs to be performed via re-crate.
+func (d *VrrpDescriptor) UpdateWithRecreate(_ string, oldVRRPEntry, newVRRPEntry *l3.VRRPEntry, _ interface{}) bool {
+	if oldVRRPEntry.Enabled == newVRRPEntry.Enabled {
+		// Something has changed, but it is not about enabled/disabled.
+		return true
+	}
+
+	return !allFieldsWhithoutEnabledEquals(oldVRRPEntry, newVRRPEntry)
+}
+
+// Update updates VPP VRRP entry.
+func (d *VrrpDescriptor) Update(_ string, oldVRRPEntry, newVRRPEntry *l3.VRRPEntry, _ interface{}) (
+	_ interface{}, err error) {
+
+	if newVRRPEntry.Enabled {
+		err = d.vrrpHandler.VppStartVrrp(newVRRPEntry)
+	} else {
+		err = d.vrrpHandler.VppStopVrrp(newVRRPEntry)
+	}
+	return nil, err
+}
+
+// Dependencies lists dependencies for a VPP VRRP entry.
+func (d *VrrpDescriptor) Dependencies(key string, vrrp *l3.VRRPEntry) (deps []kvs.Dependency) {
+	// the interface must exist
+	if vrrp.Interface != "" {
+		deps = append(deps, kvs.Dependency{
+			Label: vrrpEntryInterfaceDep,
+			Key:   interfaces.InterfaceKey(vrrp.Interface),
+		})
+	}
+	return deps
+}
+
+// Retrieve returns all configured VPP VRRP entries.
+func (d *VrrpDescriptor) Retrieve(correlate []adapter.VRRPEntryKVWithMetadata) (
+	retrieved []adapter.VRRPEntryKVWithMetadata, err error,
+) {
+	entries, err := d.vrrpHandler.DumpVrrpEntries()
+
+	for _, entry := range entries {
+		retrieved = append(retrieved, adapter.VRRPEntryKVWithMetadata{
+			Key:    l3.VrrpEntryKey(entry.Interface, entry.VrId),
+			Value:  entry,
+			Origin: kvs.UnknownOrigin,
```

Comment:
Here you can do `Origin: kvs.FromNB`
Unknown is used when VPP creates some more items automatically on its own and it is impossible to distinguish between them and items configured by descriptor. For example, VPP populates routing table with some special routes already when interface with IP address is created (even if no routes are configured explicitly).
But VRRP I assume never configures anything implicitly.